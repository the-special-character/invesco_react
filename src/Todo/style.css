.container {
  height: 100vh;
  max-width: 1024px;
  margin: 0 auto;
  display: flex;
  flex-direction: column;
  align-items: center;
}

.screenWidth {
  width: 100%;
}

.todoItem {
  display: flex;
  align-items: center;
  gap: 0.5rem;
}

.todoItem > p {
  flex: 1;
  padding: 0 1rem;
}
