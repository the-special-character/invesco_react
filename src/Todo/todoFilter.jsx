import React, { memo } from 'react';
import PropTypes from 'prop-types';

function TodoFilter({ filterTodo, filterType }) {
  return (
    <div className="screenWidth flex">
      <button
        className="flex-1"
        type="button"
        name="all"
        style={{
          backgroundColor:
            filterType === 'all' ? 'dodgerblue' : 'initial',
        }}
        onClick={filterTodo}
      >
        All
      </button>
      <button
        className="flex-1"
        type="button"
        name="pending"
        style={{
          backgroundColor:
            filterType === 'pending'
              ? 'dodgerblue'
              : 'initial',
        }}
        onClick={filterTodo}
      >
        Pending
      </button>
      <button
        className="flex-1"
        type="button"
        name="completed"
        onClick={filterTodo}
        style={{
          backgroundColor:
            filterType === 'completed'
              ? 'dodgerblue'
              : 'initial',
        }}
      >
        Completed
      </button>
    </div>
  );
}

TodoFilter.propTypes = {
  filterTodo: PropTypes.func.isRequired,
  filterType: PropTypes.oneOf([
    'all',
    'pending',
    'completed',
  ]).isRequired,
};

export default memo(TodoFilter);
