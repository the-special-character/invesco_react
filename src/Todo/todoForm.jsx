import React, { forwardRef, memo } from 'react';
import PropTypes from 'prop-types';

const TodoForm = forwardRef(({ addTodo }, ref) => {
  console.log('TodoForm');
  return (
    <form onSubmit={addTodo}>
      <input
        data-testid="todo_text"
        ref={ref}
        type="text"
        placeholder="Write your todo here..."
      />
      <button type="submit">Add Todo</button>
    </form>
  );
});

TodoForm.propTypes = {
  addTodo: PropTypes.func.isRequired,
};

export default memo(TodoForm);
