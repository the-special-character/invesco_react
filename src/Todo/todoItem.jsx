import React, { memo } from 'react';
import PropTypes from 'prop-types';

function TodoItem({
  item,
  toggleCompleteTodo,
  deleteTodo,
}) {
  return (
    <div
      data-testid="todo-item"
      key={item.id}
      className="todoItem flex"
    >
      <input
        type="checkbox"
        checked={item.isDone}
        onChange={() => toggleCompleteTodo(item)}
      />
      <p className="flex-1">{item.text}</p>
      <button
        type="button"
        onClick={() => deleteTodo(item)}
      >
        Delete
      </button>
    </div>
  );
}

TodoItem.propTypes = {
  deleteTodo: PropTypes.func.isRequired,
  toggleCompleteTodo: PropTypes.func.isRequired,
  item: PropTypes.exact({
    id: PropTypes.number.isRequired,
    text: PropTypes.string.isRequired,
    isDone: PropTypes.bool.isRequired,
  }).isRequired,
};

export default memo(TodoItem);
