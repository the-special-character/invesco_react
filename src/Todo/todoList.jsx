import React, { memo } from 'react';
import PropTypes from 'prop-types';
import TodoItem from './todoItem';

function TodoList({
  todoList,
  filterType,
  toggleCompleteTodo,
  deleteTodo,
}) {
  return (
    <div className="screenWidth flex-1">
      {todoList.map(item => {
        if (
          filterType === 'all' ||
          (item.isDone && filterType === 'completed') ||
          (!item.isDone && filterType === 'pending')
        ) {
          return (
            <TodoItem
              key={item.id}
              item={item}
              toggleCompleteTodo={toggleCompleteTodo}
              deleteTodo={deleteTodo}
            />
          );
        }
        return null;
      })}
    </div>
  );
}

TodoList.propTypes = {
  deleteTodo: PropTypes.func.isRequired,
  toggleCompleteTodo: PropTypes.func.isRequired,
  filterType: PropTypes.oneOf([
    'all',
    'pending',
    'completed',
  ]).isRequired,
  todoList: PropTypes.arrayOf(
    PropTypes.exact({
      id: PropTypes.string.isRequired,
      text: PropTypes.string.isRequired,
      isDone: PropTypes.bool.isRequired,
    }).isRequired,
  ).isRequired,
};

export default memo(TodoList);
