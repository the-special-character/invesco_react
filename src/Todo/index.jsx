import React, { Component, createRef } from 'react';
import { v4 as uuidv4 } from 'uuid';
import './style.css';
import TodoFilter from './todoFilter';
import TodoForm from './todoForm';
import TodoList from './todoList';

export default class Todo extends Component {
  constructor(props) {
    super(props);
    this.state = {
      todoList: [],
      filterType: 'all',
    };
    this.todoTextRef = createRef();
  }

  addTodo = event => {
    event.preventDefault();

    this.setState(
      ({ todoList }) => ({
        todoList: [
          ...todoList,
          {
            id: uuidv4(),
            text: this.todoTextRef.current.value,
            isDone: false,
          },
        ],
      }),
      () => {
        this.todoTextRef.current.value = '';
      },
    );
  };

  toggleCompleteTodo = item => {
    this.setState(({ todoList }) => {
      const index = todoList.findIndex(
        x => x.id === item.id,
      );

      return {
        todoList: [
          ...todoList.slice(0, index),
          {
            ...todoList[index],
            isDone: !todoList[index].isDone,
          },
          ...todoList.slice(index + 1),
        ],
      };
    });
  };

  deleteTodo = item => {
    this.setState(({ todoList }) => {
      const index = todoList.findIndex(
        x => x.id === item.id,
      );

      return {
        todoList: [
          ...todoList.slice(0, index),
          ...todoList.slice(index + 1),
        ],
      };
    });
  };

  filterTodo = event => {
    this.setState({ filterType: event.target.name });
  };

  render() {
    const { todoList, filterType } = this.state;

    return (
      <div className="container flex">
        <h1>Todo List</h1>
        <TodoForm
          addTodo={this.addTodo}
          ref={this.todoTextRef}
        />
        <TodoList
          todoList={todoList}
          filterType={filterType}
          toggleCompleteTodo={this.toggleCompleteTodo}
          deleteTodo={this.deleteTodo}
        />
        <TodoFilter filterTodo={this.filterTodo} />
      </div>
    );
  }
}
