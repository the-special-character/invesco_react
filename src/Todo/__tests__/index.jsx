import React from 'react';
import { render } from '@testing-library/react';
import Todo from '..';

jest.mock('uuid', () => ({
  v4: () => new Date().valueOf(),
}));

describe('Todo Component', () => {
  test('should render todo component', () => {
    const { container } = render(<Todo />);
    console.log(container.childNodes);
    expect(container).toBeInTheDocument();
  });
});
