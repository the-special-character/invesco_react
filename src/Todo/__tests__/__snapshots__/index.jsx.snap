// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo Component should take snapshot 1`] = `
NodeList [
  <div
    class="container flex"
  >
    <h1>
      Todo List App
    </h1>
    <form>
      <input
        data-testid="todo_text"
        placeholder="Write your todo here..."
        type="text"
      />
      <button
        type="submit"
      >
        Add Todo
      </button>
    </form>
    <div
      class="screenWidth flex-1"
    />
    <div
      class="screenWidth flex"
    >
      <button
        class="flex-1"
        name="all"
        style="background-color: dodgerblue;"
        type="button"
      >
        All
      </button>
      <button
        class="flex-1"
        name="pending"
        type="button"
      >
        Pending
      </button>
      <button
        class="flex-1"
        name="completed"
        type="button"
      >
        Completed
      </button>
    </div>
  </div>,
]
`;
